sklearn/preprocessing/tests/test_knn_imputation.py

```diff
@@ -0,0 +1,220 @@
+from __future__ import division
+import numpy as np
+
+from sklearn.utils.testing import assert_equal
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_array_almost_equal
+from sklearn.utils.testing import assert_raises
+from sklearn.utils.testing import assert_false
+
+from sklearn.preprocessing.knn_imputation import KNNImputer
+from sklearn.random_projection import sparse_random_matrix
+
+
+def test_knn_imputation_shape():
+    # Verify the shapes of the imputed matrix for different weights and
+    # number of neighbors.
+    n_rows = 10
+    n_cols = 2
+    X = np.random.rand(n_rows, n_cols)
+    X[0, 0] = np.nan
+
+    for weights in ['uniform', 'distance']:
+        for n_neighbors in range(1, 6):
+            imputer = KNNImputer(n_neighbors=n_neighbors, weights=weights)
+            X_imputed = imputer.fit_transform(X)
+            assert_equal(X_imputed.shape, (n_rows, n_cols))
+
+
+def test_knn_imputation_zero():
+    # Test imputation when missing_values == 0
+    missing_values = 0
+    n_neighbors = 2
+    imputer = KNNImputer(missing_values=missing_values,
+                         n_neighbors=n_neighbors,
+                         weights="uniform")
+
+    # Test with missing_values=0 when NaN present
+    X = np.array([
+        [np.nan, 0, 0, 0, 5],
+        [np.nan, 1, 0, np.nan, 3],
+        [np.nan, 2, 0, 0, 0],
+        [np.nan, 6, 0, 5, 13],
+    ])
+    assert_raises(ValueError, imputer.fit, X)
+
+    # Test with % zeros in column > col_max_missing
+    X = np.array([
+        [1, 0, 0, 0, 5],
+        [2, 1, 0, 2, 3],
+        [3, 2, 0, 0, 0],
+        [4, 6, 0, 5, 13],
+    ])
+    assert_raises(ValueError, imputer.fit, X)
+
+    # Test with an imputable matrix
+    X = np.array([
+        [1, 0, 1, 0, 5],
+        [2, 1, 2, 2, 3],
+        [3, 2, 3, 0, 0],
+        [6, 6, 0, 5, 13],
+    ])
+
+    statistics_mean = [3, 3, 2, 3.5, 7]
+    X_imputed = np.array([
+        [1, 1.5, 1,   2, 5],
+        [2, 1,   2,   2, 3],
+        [3, 2,   3,   2, 4],
+        [6, 6,   1.5, 5, 13],
+    ])
+
+    assert_array_equal(imputer.fit(X).transform(X), X_imputed)
+    assert_array_equal(imputer.statistics_[1], statistics_mean)
+
+
+def test_knn_imputation_default():
+    # Test imputation with default values
```

Comment:
You need more testing of each parameter to ensure that boundary cases and the backoff logic are handled correctly. If I were you, I'd make things very clear by testing on a one-column dataset, and varying the relevant parameters (`n_neighbors`, `metric`, `missing_values`, `weights`...) one by one, even in a separate test function for each; then you can check that the overall imputation is a concatenation of the one-column results, except for the `row_max_missing` behaviour which needs to be tested separately. You should also test that `KNNImputer(n_neighbors=n).fit_transform(X) == KNNImputer(n_neighbors=n + 1).fit(X).transform(X)` for multiple `n`.